### PR TITLE
OBJ 247 Update status and timestamp in mongo if processing fails

### PR DESCRIPTION
**Pull request by ICoxCH**

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java [hunk 1]
```diff
@@ -681,4 +681,4 @@ void willCallFileTransferApiForDownload() throws ServiceException {
         assertNotNull(downloadServiceResult);
         assertEquals(HttpStatus.OK, downloadServiceResult.getHttpStatus());
     }
-}
+}
```
Comment (mattch1) [inline src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java:684]: Newline

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessorTest.java [hunk 2]
```diff
@@ -192,8 +217,10 @@ void processHandlesChipsUncheckedException() throws ServiceException {
...
 
-        assertThrows(RuntimeException.class, () -> objectionProcessor.process(dummyObjection, HTTP_REQUEST_ID));
+        assertThrows(RuntimeException.class, () -> objectionProcessor.process(dummyObjection, HTTP_REQUEST_ID, objectionPatch));
 
         verify(apiLogger, times(1)).errorContext(eq(HTTP_REQUEST_ID), any(), any(), any());
+
+        assertEquals(ObjectionStatus.PROCESSING, objectionPatch.getStatus());
     }
-}
+}
```
Comment (mattch1) [inline src/test/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessorTest.java:226]: Newline

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java [hunk 3]
```diff
@@ -163,7 +163,18 @@ public void patchObjection(String objectionId,
 
         // if changing status to SUBMITTED from OPEN, process the objection
         if (ObjectionStatus.SUBMITTED == objectionPatch.getStatus() && ObjectionStatus.OPEN == previousStatus) {
-            objectionProcessor.process(objection, requestId);
+            ObjectionPatch patch = new ObjectionPatch();
+
+            try {
+                objectionProcessor.process(objection, requestId, patch);
+            } catch (Exception e) {
+                logger.infoContext(requestId, "Error processing objection updating status");
```
Comment (mattch1) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:171]: Should we log using errorContext here?

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessor.java [hunk 4]
```diff
@@ -53,9 +55,10 @@ public ObjectionProcessor(IEmailService emailService,
      *
      * @param objection     the objection to process
      * @param httpRequestId http request id used for logging
+     * @param out an ObjectionPatch that is updated as the objection is processed
      * @throws InvalidObjectionStatusException if the Objection is not currently in status SUBMITTED when this is called
      */
-    public void process(Objection objection, String httpRequestId)
+    public void process(Objection objection, String httpRequestId, ObjectionPatch out)
```
Comment (mattch1) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessor.java:61]: Could we rename the patch from "out" to something more descriptive, "such as outgongPatch" or "outgoingPatch".

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/ObjectionStatus.java [hunk 5]
```diff
@@ -6,6 +6,9 @@
     INELIGIBLE_NO_DISSOLUTION_ACTION,
     OPEN,
     PROCESSED,
+    PROCESSING,
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/ObjectionStatus.java:9]: The status in the story are ERROR_... to be set when the exception is caught. So when chips fails, the sendObjectionToChips method will set the status to ERROR_CHIPS, likewise for sendInternalEmail and sendExternalEmail. Keep 'Processing' though, that might be useful.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/patch/ObjectionPatch.java [hunk 6]
```diff
@@ -2,9 +2,12 @@
 
 import uk.gov.companieshouse.api.strikeoffobjections.model.entity.ObjectionStatus;
 
+import java.time.LocalDateTime;
+
 public class ObjectionPatch {
     private String reason;
     private ObjectionStatus status;
+    private LocalDateTime createdOn;
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/patch/ObjectionPatch.java:10]: could we just have the patch method set the current timestamp, just thinking that we could set a status with last years date if we wanted

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java (hunk 3 above)
Comment (ICoxCH) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:171]: Normally I would agree with you, but as the exception gets rethrown directly below and gets logged in the controller, that would result in duplicating the exception in the logs.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessor.java [hunk 7]
```diff
@@ -73,21 +76,25 @@ public void process(Objection objection, String httpRequestId)
 
         validateObjectionStatus(objection, httpRequestId);
 
-        // TODO update status to PROCESSING
+        out.setStatus(ObjectionStatus.PROCESSING);
+        out.setCreatedOn(LocalDateTime.now());
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessor.java:80]: If the idea is to update the status and timestamp as the code progresses, don't you have to save the entity to Mongo at each of these steps?

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessor.java [hunk 8]
```diff
@@ -142,7 +143,16 @@ private void sendExternalEmail(Objection objection, CompanyProfileApi companyPro
...
 
+            updateObjectionStatus(objection, httpRequestId, ObjectionStatus.ERROR_EXTERNAL_EMAIL);
             throw e;
         }
     }
+
+    private void updateObjectionStatus(Objection objection, String requestId, ObjectionStatus newStatus) {
+        objection.setStatus(newStatus);
+        objection.setHttpRequestId(requestId);
+        objection.setCreatedOn(LocalDateTime.now());
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/processor/ObjectionProcessor.java:154]: createdOn is set only when the objection is created, we should have a new field for the status timestamp - maybe statusChangedOn? this records when the status changed.

**Review by markpit — APPROVED** [review]

**Review by mwestacott — APPROVED** [review]
Looks good to me, but worth checking with @mattch1 too as he had earlier comments.

**Review by mattch1 — APPROVED** [review]

**Pull request merged**